adjust_hue now supports inputs of type Tensor

CristianManta opened the pull request:
Attempts to solve the bug reported in issue #2563. 

Fixes #2563 

Blocked by #2586

vfdev-5 commented:
@CristianManta thanks for the PR, the main point of unifying inputs of `F.adjust_*` methods is to ensure that the result is very close to PIL's one. See for example [here](https://github.com/pytorch/vision/blob/master/test/test_functional_tensor.py#L129) for `adjust_saturation`, `adjust_brightness` etc. I also started working on this feature but not yet sent a PR, if you would like to make your PR landed I can guide you what should be done exactly to have it merged.

CristianManta commented:
@vfdev-5 I would be happy to help on this feature if it's within my ability. I can give it a crack.

vfdev-5 commented:
@CristianManta sounds great ! Could you please confirm if you could get it done by this week ?
If so, the next step is to update the test: https://github.com/pytorch/vision/blob/master/test/test_functional_tensor.py#L129
There are two problems if we do it straight forward without any adaptations:
- https://github.com/pytorch/vision/blob/a75fdd4180683f7953d97ebbcc92d24682690f96/test/test_functional_tensor.py#L148 
range of factor is larger than expected by `F.adjust_hue`. However, the tests still fail below :
https://github.com/pytorch/vision/blob/a75fdd4180683f7953d97ebbcc92d24682690f96/test/test_functional_tensor.py#L166

- We have to verify PIL and Tensor outputs on a small image and adapt assertions accordingly...

CristianManta commented:
@vfdev-5 I can spend around 2 hours per day on it (including reading about torch.jit.script and other needed stuff) since I have a full-time summer job. I'm also using Python since only 7 months and have a lot to learn/read, so I would say that I would try it tonight and if I'm too lost or feel that I won't be able to complete it by the end of this week, I'll let you know (tonight, Montreal time) so that you can delegate this task to someone else.

So, if I understand correctly, I have to add a few tests in `test_adjustments` to make sure that the difference between `F.adjust_hue` and `F_t.adjust_hue` and the script version is not too big (in a similar fashion to the code you quoted, but with a different factor)? Basically I'll add `(F.adjust_hue, F_t.adjust_hue, script_adjust_hue)` at line 134 and see how to make the test pass, right?

I'll read about `torch.jit.script` also, since I never used it before.

vfdev-5 commented:
> @vfdev-5 I can spend around 2 hours per day on it (including reading about torch.jit.script and other needed stuff) since I have a full-time summer job. I'm also using Python since only 7 months and have a lot to learn/read, so I would say that I would try it tonight and if I'm too lost or feel that I won't be able to complete it by the end of this week, I'll let you know (tonight, Montreal time) so that you can delegate this task to someone else.

@CristianManta no worries, you can start it and I can finish your PR by pushing to your branch.

> So, if I understand correctly, I have to add a few tests in test_adjustments to make sure that the difference between F.adjust_hue and F_t.adjust_hue and the script version is not too big (in a similar fashion to the code you quoted, but with a different factor)? Basically I'll add (F.adjust_hue, F_t.adjust_hue, script_adjust_hue) at line 134 and see how to make the test pass, right?

yes, and I know that it wont pass as I checked that locally. That's why 
> We have to verify PIL and Tensor outputs on a small image and adapt assertions accordingly...
(see my above message).

> I'll read about torch.jit.script also, since I never used it before.

There is nothing complicated here, mostly the task is to fixed runtime errors while scripting the function and its subfunctions as probably several variables do not have explicitly declare their types.

CristianManta commented:
@vfdev-5 I added the appropriate assertions in the tests, but of course there is an error when defining `torch.jit.script(F_t.adjust_hue)` since I guess I need to modify `F_t.adjust_hue` to make it scriptable. However, I read the tutorials and examples about converting functions to scripts and I can reproduce them on trivial examples like 

```python
import torch                                                                                                                                                                                  
@torch.jit.script                                                               
def func(x):                                                                    
    return x**2                                                                 
                                                                                
def func2(z):                                                                   
    return 2*func(2*z)                                                          
print(func.code)                                                                
                                                                                
sc2 = torch.jit.script(func2)                                                   
print(sc2.code) 
```
so I'm not sure to understand why it doesn't work with `F_t.adjust_hue`. I tried adding type declarations (that you can see in my updated PR) since you mentioned the lack of type declarations. Am I supposed to add a type in front of each variable involved in the function to make it work (I doubt that, because the errors were at the same place regardless of if I added those types or not)?

vfdev-5 commented:
@CristianManta sorry for delay, yes, you need to do exactly as it is done for example for [`F_t.adjust_gamma`](https://github.com/pytorch/vision/blob/master/torchvision/transforms/functional_tensor.py#L201). 
Another thing to do is to add type hints to `_hsv2rgb` and `_rgb2hsv` functions. In my tests we also have to modify the implementation of `_rgb2hsv`:
```python
-    s = cr / torch.where(eqc, maxc.new_ones(()), maxc)
+    ones = torch.ones_like(maxc)
+    s = cr / torch.where(eqc, ones, maxc)

-    cr_divisor = torch.where(eqc, maxc.new_ones(()), cr)
+    cr_divisor = torch.where(eqc, ones, cr)
```
Feel free to ask if you have other questions.

vfdev-5 commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -164,7 +163,7 @@ def adjust_hue(img, hue_factor):
         img = img.to(dtype=torch.float32) / 255.0
 
-    img = _rgb2hsv(img)
+    img: Tensor = _rgb2hsv(img)
```
I think this is not necessary to add `: Tensor`

vfdev-5 commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -362,7 +361,7 @@ def _rgb2hsv(img):
     cr = maxc - minc
     # Since `eqc => cr = 0`, replacing denominator with 1 when `eqc` is fine.
-    s = cr / torch.where(eqc, maxc.new_ones(()), maxc)
+    s: Tensor = cr / torch.where(eqc, maxc.new_ones(()), maxc)
```
Same here

CristianManta commented:
Ahh I see, yes I recall that the test generated an error coming from `_rgb2hsv` and I was turning around in circles to try to solve it. I'll continue working on it this evening and I'll add the remaining type hints in the functions headers that are being called. I think I'll also add a `raise TypeError` when the input does not match the type hint of the function arguments.

CristianManta commented on torchvision/transforms/functional.py:
```diff
@@ -714,7 +714,7 @@ def adjust_saturation(img: Tensor, saturation_factor: float) -> Tensor:
     return F_t.adjust_saturation(img, saturation_factor)
 
 
-def adjust_hue(img: Tensor, hue_factor: float) -> Tensor:
```
Technically speaking, the type of `hue_factor` is `torch.Tensor` of `dtype=whatever_torch.rand()'s_output_type_is`, but I guess it would be misleading to put `hue_factor: Tensor` in the header

CristianManta replied:
Also, since we're in `functional.py`, the type of `img` would be something like `Union[Tensor, PIL]`, but I cannot use `PIL` as a type so I left `img` as it is. Not sure what to do. I don't think putting `Any` is correct either.

vfdev-5 replied:
Here the type hints are mostly for `torch.jit.script` which does not accept type mixing with `Union`. Let's keep annotations as it is done for other. This is correct`def adjust_hue(img: Tensor, hue_factor: float) -> Tensor:`

CristianManta commented on torchvision/transforms/functional.py:
```diff
@@ -729,20 +729,20 @@ def adjust_hue(img: Tensor, hue_factor: float) -> Tensor:
+    return F_t.adjust_hue(img, hue_factor)
 
 
 def adjust_gamma(img: Tensor, gamma: float, gain: float = 1) -> Tensor:
```
BTW, I looked into the newest commit in master regarding `adjust_gamma` and its header still contains `img: Tensor` and `-> Tensor`, which is a mistake since in `functional.py` these methods accept both `Tensor` and `PIL` types (same remark with the return type).

CristianManta replied:
I didn't touch it to not create conflicts

vfdev-5 replied:
As explained above this is a corrected annotation for working with jit. We shouldn't change this.

CristianManta commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -129,7 +129,7 @@ def adjust_contrast(img: Tensor, contrast_factor: float) -> Tensor:
 
 
-def adjust_hue(img, hue_factor):
+def adjust_hue(img: Tensor, hue_factor: float) -> Tensor:
```
Forgot to add an `isinstance` check for the type of `img` to be `torch.Tensor`. Will do it in the next commit.

vfdev-5 replied:
Yes, please, something like here: https://github.com/pytorch/vision/blob/270e02ecaf6341269f27523b92c1b72d117aacf4/torchvision/transforms/functional_tensor.py#L635

CristianManta commented:
I think I covered all the type hint problems and I adapted the test to take into account the different range of `factor` for `adjust_hue`. The `torch.jit.script` also seems to work. I also added some comments/remarks to what I just pushed.

I guess the remaining part is to fix the error bound in the `assertLess` which is now the only source of failure in the tests. I temporarily set it to 5 just to make sure that there are no other issues.

Now, in order to fix it, I'm wondering what approach I should follow. The `5 / 255` factor seems to come from the worst case when there is a conversion from `uint8` to `torch.float` which can be off by at most 1 (rounding to nearest integer error) before dividing by 255, hence an error of `1 / 255` for  one such conversion. But why 5 times? Also, where does the `1e-5` factor come from?

If all of this is correct, then isn't changing the assertion cheating? Shouldn't we rather change `F_t.adjust_hue` to make it more precise?

vfdev-5 commented on test/test_functional_tensor.py:
```diff
@@ -146,25 +148,36 @@ def test_adjustments(self):
                 img = torch.randint(0, 256, shape, dtype=torch.uint8)
 
             factor = 3 * torch.rand(1)
```
As factor should be a float for all ops, let's just apply `.item()` on it.

vfdev-5 commented on test/test_functional_tensor.py:
```diff
@@ -146,25 +148,36 @@ def test_adjustments(self):
-                img_pil = transforms.ToPILImage()(img)
-                f_img_pil = f(img_pil, factor)
-                f_img = transforms.ToTensor()(f_img_pil)
+            for i, (f, ft, sft) in enumerate(fns):
```
Maybe it would be more simple to check the type of `f` like `f == F.adjust_hue` and redefine factor as `factor = torch.rand(1).item() - 0.5` instead of doing `if-else` with almost the same code for both cases...

vfdev-5 commented on torchvision/transforms/functional_pil.py:
```diff
@@ -118,7 +118,7 @@ def adjust_saturation(img, saturation_factor):
 
 @torch.jit.unused
-def adjust_hue(img, hue_factor):
+def adjust_hue(img, hue_factor: float):
```
This is not necessary. Otherwise, you can set it as `def adjust_hue(img: Any, hue_factor: float):`

vfdev-5 commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -344,7 +344,9 @@ def _blend(img1: Tensor, img2: Tensor, ratio: float) -> Tensor:
 
-def _rgb2hsv(img):
+def _rgb2hsv(img: Tensor) -> Tensor:
+    if not isinstance(img, torch.Tensor):
```
`_rgb2hsv` is a private method, I think we can skip type checking and raising error here.
Btw, `functional_tensor.py` module is intended to be private too, user should not use those functions in the code. Warnings will be added soon according to https://github.com/pytorch/vision/issues/2547

vfdev-5 commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -380,7 +383,9 @@ def _rgb2hsv(img):
 
-def _hsv2rgb(img):
+def _hsv2rgb(img: Tensor) -> Tensor:
+    if not isinstance(img, torch.Tensor):
```
Same here

vfdev-5 commented on test/test_functional_tensor.py:
```diff
@@ -146,25 +148,36 @@ def test_adjustments(self):
                 max_diff_scripted = (sft_img - f_img).abs().max()
-                self.assertLess(max_diff, 5 / 255 + 1e-5)
-                self.assertLess(max_diff_scripted, 5 / 255 + 1e-5)
+                self.assertLess(max_diff, 5 + 1e-5)  # TODO: 5 / 255, not 5
```
Maybe, we can compare number of different pixels like it is done here: https://github.com/pytorch/vision/blob/270e02ecaf6341269f27523b92c1b72d117aacf4/test/test_functional_tensor.py#L587

vfdev-5 commented:
> Now, in order to fix it, I'm wondering what approach I should follow. The 5 / 255 factor seems to come from the worst case when there is a conversion from uint8 to torch.float which can be off by at most 1 (floor/ceiling rounding error) before dividing by 255, hence an error of 1 / 255 for one such conversion. But why 5 times? Also, where does the 1e-5 factor come from?
> If all of this is correct, then isn't changing the assertion cheating? Shouldn't we rather change F_t.adjust_hue to make it more precise?

I was thinking, firstly, to see how different two outputs (PIL and Tensor) are, do we have majority of exact values and few different values or they are all different ? If you could compute `F.adjust_hue` on a small image and tensor of size 5x5 (e.g. `torch.arange(3 * 5 * 5).reshape(3, 5, 5)` and post here few values, it would be nice.

Next, we can take a look at `F_t.adjust_hue` implementation and see if it could be slightly changed to match PIL result.

Finally, we adapt the test such that either we compute the number of different pixels and raise an error if above some threshold. Either, rework current checking of max diff and leave comment what we check.

CristianManta commented:
That's a good idea, I'll compute the distances between pairs of images in the 2 norms you mentioned and this should give us a clue more about what's wrong. My suspicions is that many pixels will be different, slightly exceeding the established max difference each.

I'll compute `F.adjust_hue` on a PIL image, convert to Tensor and re-compute `F.adjust_hue` on a Tensor (representing the same image) and measure the distances between the 2 outputs. I'll repeat that a few times.

CristianManta commented:
I fixed the mistakes that you mentioned in the previous review. I wrote a script to get some statistics about the error differences. If you want to run it, you need to place it one level above the `vision` directory:

```python
import torch
import vision.torchvision.transforms.functional as F
import vision.torchvision.transforms.transforms as transforms
import vision.torchvision.transforms.functional_tensor as F_t


# Parameters
tol = 1e-5
k = 20

def get_number_of_diff_pixels(img1, img2):
    W = img1.shape[1]
    H = img1.shape[2]
    counter = 0
    for i in range(W):
        for j in range(H):
            for ch in range(3):
                if abs(img1[ch][i][j].item() - img2[ch][i][j].item()) > tol:
                    counter += 1
                    print('The discrepancy was in channel: {}'.format(ch))
                    break
    return counter


tensor_img = torch.arange(3 * 5 * 5).reshape(3, 5, 5).to(torch.float) / 255
pil_img = transforms.ToPILImage()(tensor_img)

# first testing if composing the ToPILImage and its inverse produces the 
# original image indeed

test_inverse_pil_img = transforms.ToTensor()(pil_img)

print("0 means ok: ", (test_inverse_pil_img != tensor_img).sum().item() / 3.0)
print('\n')

for i in range(k + 1):
    hue_factor = -0.5 + i / k
    out_tensor = F.adjust_hue(tensor_img, hue_factor)
    out_pil = F.adjust_hue(pil_img, hue_factor)
    out_pil_to_tensor = transforms.ToTensor()(out_pil)

    # num_diff_pixels = (out_pil_to_tensor != out_tensor).sum().item() / 3.0
    num_diff_pixels = get_number_of_diff_pixels(out_pil_to_tensor, out_tensor)
    print('i = {}: number of different pixels = {}'.format(i, num_diff_pixels))
    pct = 100.0 * num_diff_pixels / (tensor_img.shape[1] * tensor_img.shape[2])
    print('percentage of pixels that are different: {}%\n'.format(pct))

# testing if _rgb2hsv and _hsv2rgb are precise:
hsv = F_t._rgb2hsv(tensor_img)
rgb = F_t._hsv2rgb(hsv)

diff = get_number_of_diff_pixels(rgb, tensor_img)
print('diff = {}'.format(diff))
```

On some relatively mild tolerance `tol`, on most `hue_factor` values tried, all 25 pixels differed by more than `tol` when comparing the original tensor to the one got using the PIL transformations. For instance, if you set `tol = 1e-5` (even `1e-4`), it looks pretty bad. In particular, when `i = 10`, `hue_factor = 0` and the image is supposed to be left unchanged. Yet, all 25 pixels differ. The problem doesn't seem to favour one channel in particular.

I also tested to see if the 2 conversions `_rgb2hsv` and its inverse have bad accuracy by composing one by the other and comparing with the original tensor. A difference can already be noticed with `tol = 1e-7`.

Sorry if the output of the script isn't pretty. I didn't have time to make it prettier

vfdev-5 commented on test/test_functional_tensor.py:
```diff
@@ -163,20 +168,24 @@ def test_adjustments(self):
             scripted_fn(img)
 
-            f = transforms.ColorJitter(saturation=factor.item())
+            f = transforms.ColorJitter(hue=abs(hue_factor))
```
Why do you use abs here ?

CristianManta replied:
Because the `ColorJitter` class only takes `hue` inputs that are non negative: https://github.com/pytorch/vision/blob/c547e5cbbf25e46870abb2e925b42c67c3367afc/torchvision/transforms/transforms.py#L1024

CristianManta replied:
I also had to create 2 new variables, `bcs_factor` and `hue_factor` because after the inner loop, once `factor` has been overwritten, I still needed the original `factor` and `hue_factor`

vfdev-5 commented:
@CristianManta I was working on the test of another color transform: `rgb_to_grayscale` and as some of `adjust_*` uses it I think it would be better to refactor this `test_adjustments` into separate tests and follow the same test schema as for `test_adjust_gamma`. Here is my working branch: https://github.com/pytorch/vision/commit/d016cabc5f40654dc217e83c749b7f7a1f9594e0
I think it would be better to rewrite the test of `adjust_hue` in the same way.

CristianManta commented:
@vfdev-5 I copied your `test_functional_tensor.py` file from your working branch to make my editing simpler and to make it consistent with your branch and I added `test_adjust_hue` following the same style and `test_class_interface` since it was tested in master but you removed it (why?)

There is a conflict which seems trivial to resolve by just removing the duplicate fragment, but I think I'll let you do it. I'm not sure why that happened

vfdev-5 commented:
> There is a conflict which seems trivial to resolve by just removing the duplicate fragment. With your permission, I can resolve it here on GitHub. I'm not sure why that happened.

@CristianManta , yes please. As it happens on your branch you can do whatever you want without any permission :)

 >  test_class_interface since it was tested in master but you removed it (why?)

Well, technically, all class interfaces have to be tested in https://github.com/pytorch/vision/blob/master/test/test_transforms_tensor.py
it would be better to test `ColorJitter` there too.

CristianManta commented:
Ok, so can I also remove `test_adjustments` from there since we already do those tests in `test_functional_tensor.py`?

vfdev-5 commented:
Yes, you can remove `test_adjustments` from `test_transforms_tensor.py`, but we need to add a test for ColorJitter with all 4 options: brightness, contrast, saturation and hue.

CristianManta commented:
@vfdev-5 There is still one more test failing, `test_adjust_hue` reports that the difference is too big still. I tried fixing this by using `colorsys` instead of `_rgb2hsv` and `_hsv2rgb` to perhaps reduce error coming from them, but `colorsys` functions are not scriptable. What should I do?

fmassa commented:
@CristianManta I think there might be some changes that went in this PR that are unrelated (in the tests).

About the differences from `_rgb2hsv` and `_hsv2rgb`, do they also fail on master? I think I've seen some occasional failures.
In the end, our implementation should follow more closely the official definitions. PIL does some approximations for working with `uint8` images, so I'm not so worried if it fails for some input types.

fmassa commented:
@vfdev-5 could you help take this PR into completion?

vfdev-5 commented:
Travis failed: https://travis-ci.org/github/pytorch/vision/jobs/723374281 
```
/functional_tensor.py", line 197, in adjust_hue

E               img = _rgb2hsv(img)

E               h, s, v = img.unbind(0)

E               h += hue_factor

E               ~~~~~~~~~~~~~~~ <--- HERE

E               h = h % 1.0

E               img = torch.stack((h, s, v))

E           RuntimeError: Output 0 of UnbindBackward is a view and is being modified inplace. This view is the output of a function that returns multiple views. Such functions do not allow the output views to be modified inplace. You should replace the inplace operation by an out-of-place one.
```

fmassa commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -185,17 +185,16 @@ def adjust_hue(img, hue_factor):
     h, s, v = img.unbind(0)
-    h += hue_factor
-    h = h % 1.0
+    h = torch.fmod(h + hue_factor, 1.0)
```
I believe this introduces a change in behavior: [`%` is equivalent to `torch.remainder` and not `torch.fmod`](https://github.com/pytorch/pytorch/blob/4ef12be900e9ec8b81fd45b2f7fe14d648826f64/tools/autograd/templates/python_variable_methods.cpp#L1144). Just to be sure, can you revert back to using `%`?

fmassa reviewed (commented):
One more comment, otherwise looks good to me

vfdev-5 commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -185,17 +185,16 @@ def adjust_hue(img, hue_factor):
     h, s, v = img.unbind(0)
-    h += hue_factor
-    h = h % 1.0
+    h = (h + hue_factor) % 1.0
```
This change is due to (locally irreproducible, reproducible only on Travis) issue with torch.jit https://github.com/pytorch/vision/pull/2566#issuecomment-685469496

fmassa reviewed (approved):
Thanks a lot @CristianManta and @vfdev-5 !

yaox12 commented:
Hi, I think `F_t.adjust_hue` cannot process images in batchs, which is inconsistent with `F_t.adjust_saturation`, `F_t.adjust_brightness` etc.

vfdev-5 commented:
@yaox12 yes, current implementation should be improved for that case which will be covered by https://github.com/pytorch/vision/issues/2583
